File: app/references/go/extraction.go
package extraction

import (
	"regexp"
	"strconv"
	"strings"
	"time"
)

type DocumentState string

const (
	Uploaded           DocumentState = "uploaded"
	QueuedForOCR       DocumentState = "queued_for_ocr"
	OCRProcessing      DocumentState = "ocr_processing"
	OCRDone            DocumentState = "ocr_done"
	MetadataProcessing DocumentState = "metadata_processing"
	Complete           DocumentState = "complete"
	Failed             DocumentState = "failed"
)

type Input struct {
	Text             string
	OriginalFilename string
	CreatedAt        *time.Time
	ExistingTitle    string
}

var invoiceNumberPatterns = []*regexp.Regexp{
	regexp.MustCompile(`(?i)\bEingangsrechnung\s+([A-Z0-9][A-Z0-9./-]{2,})\b`),
	regexp.MustCompile(`(?i)\bRechnungs\s*nummer\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b`),
	regexp.MustCompile(`(?i)\bRechnungs\s*nr\.?\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b`),
	regexp.MustCompile(`(?i)\bRechnungs\s*-\s*Nr\.?\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b`),
	regexp.MustCompile(`(?i)\bRechnung\s*-\s*Nr\.?\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b`),
	regexp.MustCompile(`(?i)\bRechnung\s*Nr\.?\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b`),
	regexp.MustCompile(`(?i)\bInvoice\s*No\.?\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b`),
	regexp.MustCompile(`(?i)\bInvoice\s*Number\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b`),
	regexp.MustCompile(`(?i)\bBeleg(?:nr|nummer)\.?\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b`),
}

func compactParty(value string) string {
	value = regexp.MustCompile(`(?i)\b(ges\.?\s*mbh|gesmbh|gmbh|mbh|ag|kg|ug|inc|ltd|llc|sarl|e\.?k\.?)\b`).ReplaceAllString(value, " ")
	value = regexp.MustCompile(`[^A-Za-z0-9 ]+`).ReplaceAllString(value, " ")
	parts := strings.Fields(value)
	if len(parts) > 3 {
		parts = parts[:3]
	}
	out := strings.Join(parts, "")
	if out == "" {
		return "Dok"
	}
	if len(out) > 40 {
		return out[:40]
	}
	return out
}

func compactSender(value string) string {
	value = strings.ReplaceAll(value, "&", " And ")
	value = regexp.MustCompile(`[_/\\|]+`).ReplaceAllString(value, " ")
	value = regexp.MustCompile(`[^A-Za-z0-9 .,-]+`).ReplaceAllString(value, " ")
	parts := strings.Fields(value)
	if len(parts) > 3 {
		parts = parts[:3]
	}
	for i, part := range parts {
		if part == strings.ToUpper(part) && len(part) <= 8 {
			continue
		}
		if len(part) > 0 {
			parts[i] = strings.ToUpper(part[:1]) + part[1:]
		}
	}
	out := regexp.MustCompile(`[^A-Za-z0-9]+`).ReplaceAllString(strings.Join(parts, ""), "")
	if len(out) > 32 {
		return out[:32]
	}
	return out
}

func NormalizeDate(raw string) string {
	re := regexp.MustCompile(`(\d{1,2})[./-](\d{1,2})[./-](\d{2,4})`)
	m := re.FindStringSubmatch(raw)
	if m == nil {
		return ""
	}
	day, _ := strconv.Atoi(m[1])
	month, _ := strconv.Atoi(m[2])
	year, _ := strconv.Atoi(m[3])
	if year < 100 {
		if year < 70 {
			year += 2000
		} else {
			year += 1900
		}
	}
	return fmtDate(day, month, year)
}

func fmtDate(day, month, year int) string {
	return two(day) + "/" + two(month) + "/" + strconv.Itoa(year)
}

func two(n int) string {
	if n < 10 {
		return "0" + strconv.Itoa(n)
	}
	return strconv.Itoa(n)
}

func NormalizeAmount(raw string) string {
	value := regexp.MustCompile(`[^0-9,.\-]`).ReplaceAllString(strings.TrimSpace(raw), "")
	if value == "" {
		return "NA"
	}
	var candidates []string
	if strings.Contains(value, ",") && strings.Contains(value, ".") {
		if strings.LastIndex(value, ",") > strings.LastIndex(value, ".") {
			candidates = append(candidates, strings.ReplaceAll(strings.ReplaceAll(value, ".", ""), ",", "."))
		} else {
			candidates = append(candidates, strings.ReplaceAll(value, ",", ""))
		}
	} else if strings.Contains(value, ",") {
		parts := strings.Split(value, ",")
		last := parts[len(parts)-1]
		if len(last) == 2 {
			candidates = append(candidates, strings.ReplaceAll(strings.ReplaceAll(value, ".", ""), ",", "."))
		} else if len(parts) == 2 && len(last) == 3 {
			candidates = append(candidates, strings.ReplaceAll(value, ",", ""))
		}
	} else if strings.Contains(value, ".") {
		parts := strings.Split(value, ".")
		if len(parts) == 2 && len(parts[1]) == 3 {
			candidates = append(candidates, strings.ReplaceAll(value, ".", ""))
		} else {
			candidates = append(candidates, value)
		}
	} else {
		candidates = append(candidates, value)
	}
	lastSep := regexp.MustCompile(`([.,])(\d{2})$`).FindStringSubmatchIndex(value)
	if lastSep != nil {
		intPart := regexp.MustCompile(`[^0-9]`).ReplaceAllString(value[:lastSep[2]], "")
		frac := value[lastSep[4]:lastSep[5]]
		if intPart != "" {
			candidates = append(candidates, intPart+"."+frac)
		}
	}
	for _, candidate := range candidates {
		if parsed, err := strconv.ParseFloat(candidate, 64); err == nil {
			return strings.ReplaceAll(strconv.FormatFloat(parsed, 'f', 2, 64), ".", ",")
		}
	}
	return "NA"
}

func ExtractInvoiceNumber(text string) string {
	clean := regexp.MustCompile(`[^A-Za-z0-9./-]+`)
	for _, line := range strings.Split(text, "\n") {
		low := strings.ToLower(strings.TrimSpace(line))
		if strings.Contains(low, "kundennummer") || strings.Contains(low, "customer") || strings.Contains(low, "auftrags") || strings.Contains(low, "bestell") {
			continue
		}
		for _, pattern := range invoiceNumberPatterns {
			if m := pattern.FindStringSubmatch(line); m != nil {
				out := clean.ReplaceAllString(m[1], "")
				if len(out) > 40 {
					return out[:40]
				}
				return out
			}
		}
	}
	return "NA"
}

func invoiceDate(text string, createdAt *time.Time) string {
	lines := strings.Split(text, "\n")
	for _, line := range lines {
		low := strings.ToLower(line)
		if strings.Contains(low, "rechnungsdatum") || strings.Contains(low, "invoice date") {
			if d := NormalizeDate(line); d != "" {
				return d
			}
		}
	}
	for _, line := range lines {
		low := strings.ToLower(line)
		if strings.Contains(low, "lieferdatum") || strings.Contains(low, "leistungsdatum") || strings.Contains(low, "valutadatum") {
			continue
		}
		if strings.Contains(low, "datum") || strings.Contains(low, "date") {
			if d := NormalizeDate(line); d != "" {
				return d
			}
		}
	}
	if createdAt != nil {
		return fmtDate(createdAt.Day(), int(createdAt.Month()), createdAt.Year())
	}
	return "00/00/0000"
}

func invoiceAmount(text string) string {
	numRe := regexp.MustCompile(`\b\d{1,3}(?:[.,]\d{3})*(?:[.,]\d{2})\b|\b\d+[.,]\d{2}\b|\b\d+\b`)
	type row struct {
		score int
		value float64
		amount string
	}
	var rows []row
	for _, line := range strings.Split(text, "\n") {
		nums := numRe.FindAllString(line, -1)
		if len(nums) == 0 {
			continue
		}
		low := strings.ToLower(line)
		score := 0
		if regexp.MustCompile(`endsumme|gesamtsumme|gesamtbetrag|zu zahlen|balance due|invoice total|grand total|brutto`).MatchString(low) {
			score = 4
		} else if regexp.MustCompile(`summe|gesamt|total|rechnungsbetrag`).MatchString(low) {
			score = 2
		}
		if regexp.MustCompile(`netto|mwst|ust|steuer|skonto|rabatt`).MatchString(low) && score < 4 {
			score -= 3
		}
		amount := NormalizeAmount(nums[len(nums)-1])
		if amount == "NA" {
			continue
		}
		val, _ := strconv.ParseFloat(strings.ReplaceAll(amount, ",", "."), 64)
		rows = append(rows, row{score, val, amount})
	}
	best := row{score: -999}
	for _, candidate := range rows {
		if candidate.score > best.score || (candidate.score == best.score && candidate.value > best.value) {
			best = candidate
		}
	}
	if best.score == -999 {
		return "NA"
	}
	return best.amount
}

func ExtractBelegeSender(text, originalFilename string) string {
	lines := nonEmptyLines(text)
	if len(lines) >= 2 && regexp.MustCompile(`^[A-Z]{3,8}$`).MatchString(lines[0]) && regexp.MustCompile(`^[A-Z][A-Z ]{3,24}$`).MatchString(lines[1]) {
		return compactSender(lines[0])
	}
	bad := regexp.MustCompile(`(?i)worldhealthorganization|sauglinge|impf|masern|rki|who|therapie|zuzahlung|behandl|fango|microsoft|bgm`)
	for i, line := range lines {
		if i >= 12 {
			break
		}
		token := compactSender(line)
		low := strings.ToLower(token)
		if token == "" || bad.MatchString(low) || strings.Contains(line, ":") || regexp.MustCompile(`\d{2,}`).MatchString(line) {
			continue
		}
		if regexp.MustCompile(`(?i)bank|gmbh|ag|market|shop|store|distributors`).MatchString(line) || regexp.MustCompile(`^[A-Z][A-Za-z0-9]{2,20}$`).MatchString(token) {
			return token
		}
	}
	stem := regexp.MustCompile(`\.[A-Za-z0-9]{1,5}$`).ReplaceAllString(originalFilename, "")
	if stem != "" && !regexp.MustCompile(`(?i)scan|img|image|document|invoice|rechnung|impf|therapie`).MatchString(stem) {
		return compactSender(stem)
	}
	return "Dok"
}

func ExtractEingangsrechnungSender(text, originalFilename string) string {
	for i, line := range nonEmptyLines(text) {
		if i >= 18 {
			break
		}
		if regexp.MustCompile(`(?i)firma|kunde|rechnung|invoice|datum|summe|telefon|mail|www|lieferdatum`).MatchString(line) {
			continue
		}
		candidate := regexp.MustCompile(`(?i)\b(phone|telefon|fax|mail|www)\b`).Split(line, 2)[0]
		candidate = regexp.MustCompile(`(?i)\b(strasse|straße|str\.?|weg|gasse|allee|platz|ufer|ring)\b`).Split(candidate, 2)[0]
		candidate = regexp.MustCompile(`\b\d{4,5}\b`).Split(candidate, 2)[0]
		sender := compactParty(candidate)
		if sender != "Dok" && len(sender) >= 3 {
			return sender
		}
	}
	if originalFilename != "" {
		return compactParty(regexp.MustCompile(`\.[A-Za-z0-9]{1,5}$`).ReplaceAllString(originalFilename, ""))
	}
	return "Dok"
}

func ExtractAusgangsrechnungRecipient(text string) string {
	meta := regexp.MustCompile(`(?i)\b(rechnung|rechnungs|invoice|datum|date|kunden-?nr|kundennummer|bestell-?nr|auftrags-?nr|telefon|phone|fax|mail|www|summe|ust|mwst|endsumme|iban|bic|konto|pos\.)`)
	var clean []string
	seen := map[string]bool{}
	for i, line := range nonEmptyLines(text) {
		if i >= 25 || meta.MatchString(line) {
			break
		}
		if regexp.MustCompile(`@|www\.|\+\d|^\d{4,5}\s|\d|\b(?i:strasse|straße|str\.?|weg|gasse|allee|platz|ufer|ring)\b`).MatchString(line) {
			continue
		}
		candidate := compactParty(line)
		if candidate != "Dok" && len(candidate) >= 3 && !seen[candidate] {
			seen[candidate] = true
			clean = append(clean, candidate)
		}
	}
	if len(clean) >= 2 {
		return clean[1]
	}
	return "Dok"
}

func ExtractBelegeTitle(input Input) string {
	date := "00/00"
	if input.CreatedAt != nil {
		date = two(int(input.CreatedAt.Month())) + "/" + two(input.CreatedAt.Year()%100)
	}
	return ExtractBelegeSender(input.Text, input.OriginalFilename) + "_B_" + date + "_" + belegeAmount(input.Text) + "_" + payment(input.Text)
}

func ExtractEingangsrechnungTitle(input Input) string {
	return ExtractEingangsrechnungSender(input.Text, input.OriginalFilename) + "_" + ExtractInvoiceNumber(input.Text) + "_" + invoiceDate(input.Text, input.CreatedAt) + "_" + invoiceAmount(input.Text)
}

func ExtractAusgangsrechnungTitle(input Input) string {
	return ExtractAusgangsrechnungRecipient(input.Text) + "_" + ExtractInvoiceNumber(input.Text) + "_" + invoiceDate(input.Text, input.CreatedAt) + "_" + invoiceAmount(input.Text)
}

func belegeAmount(text string) string {
	numRe := regexp.MustCompile(`\b\d{1,3}(?:[.,]\d{3})*(?:[.,]\d{2})\b|\b\d+[.,]\d{2}\b|\b\d+\b`)
	bestScore := -1
	best := ""
	for _, line := range strings.Split(text, "\n") {
		score := 0
		low := strings.ToLower(line)
		if regexp.MustCompile(`summe|gesamt|total|betrag|balance due|zu zahlen|endbetrag|invoice total|rechnungsbetrag`).MatchString(low) {
			score += 2
		}
		if strings.Contains(low, "eur") || strings.Contains(line, "€") || strings.Contains(line, "$") {
			score++
		}
		for _, number := range numRe.FindAllString(line, -1) {
			if score > bestScore || (score == bestScore && len(number) > len(best)) {
				bestScore = score
				best = number
			}
		}
	}
	if bestScore < 2 {
		return "NA"
	}
	return NormalizeAmount(best)
}

func payment(text string) string {
	low := strings.ToLower(text)
	if regexp.MustCompile(`\b(bar|cash|barzahlung)\b`).MatchString(low) {
		return "Bar"
	}
	if regexp.MustCompile(`\b(karte|ec|girocard|visa|mastercard|amex|electronic cash|debit)\b`).MatchString(low) {
		return "Karte"
	}
	return "NA"
}

func nonEmptyLines(text string) []string {
	var out []string
	for _, line := range strings.Split(text, "\n") {
		line = strings.TrimSpace(line)
		if line != "" {
			out = append(out, line)
		}
	}
	return out
}

File: app/references/typescript/src/extraction.ts
export type DocumentState =
  | 'uploaded'
  | 'queued_for_ocr'
  | 'ocr_processing'
  | 'ocr_done'
  | 'metadata_processing'
  | 'complete'
  | 'failed'

export interface ExtractionInput {
  text: string
  originalFilename?: string
  createdAt?: Date
  existingTitle?: string
}

const badBelege = [
  'worldhealthorganization', 'sauglinge', 'impf', 'masern', 'rki', 'who',
  'therapie', 'zuzahlung', 'behandl', 'fango', 'microsoft', 'bgm'
]

function stripAccents(value: string): string {
  return value.normalize('NFD').replace(/\p{Diacritic}/gu, '').replaceAll('ß', 'ss')
}

function compactParty(value: string): string {
  const cleaned = stripAccents(value)
    .replace(/\b(ges\.?\s*mbh|gesmbh|gmbh|mbh|ag|kg|ug|inc|ltd|llc|sarl|e\.?k\.?)\b/gi, ' ')
    .replace(/[^A-Za-z0-9 ]+/g, ' ')
    .replace(/\s+/g, ' ')
    .trim()
  return cleaned.split(/\s+/).filter(Boolean).slice(0, 3).join('').slice(0, 40) || 'Dok'
}

function compactSender(value: string): string {
  const words = stripAccents(value)
    .replace(/&/g, ' And ')
    .replace(/[_/\\|]+/g, ' ')
    .replace(/[^A-Za-z0-9 .,-]+/g, ' ')
    .trim()
    .split(/\s+/)
    .filter(Boolean)
  const token = words.slice(0, 3).map((word) => word.toUpperCase() === word && word.length <= 8 ? word : word[0].toUpperCase() + word.slice(1)).join('')
  return token.replace(/[^A-Za-z0-9]+/g, '').slice(0, 32)
}

export function normalizeDate(raw: string): string {
  const match = raw.match(/(\d{1,2})[./-](\d{1,2})[./-](\d{2,4})/)
  if (!match) return ''
  let year = Number(match[3])
  if (year < 100) year = year < 70 ? 2000 + year : 1900 + year
  return `${Number(match[1]).toString().padStart(2, '0')}/${Number(match[2]).toString().padStart(2, '0')}/${year.toString().padStart(4, '0')}`
}

export function normalizeAmount(raw: string): string {
  const value = raw.trim().replace(/[^0-9,.-]/g, '')
  if (!value) return 'NA'
  const candidates: string[] = []
  if (value.includes(',') && value.includes('.')) {
    candidates.push(value.lastIndexOf(',') > value.lastIndexOf('.') ? value.replace(/\./g, '').replace(',', '.') : value.replace(/,/g, ''))
  } else if (value.includes(',')) {
    const parts = value.split(',')
    if (parts.at(-1)?.length === 2) candidates.push(value.replace(/\./g, '').replace(',', '.'))
    else if (parts.length === 2 && parts.at(-1)?.length === 3) candidates.push(value.replace(/,/g, ''))
  } else if (value.includes('.')) {
    const parts = value.split('.')
    candidates.push(parts.length === 2 && parts.at(-1)?.length === 3 ? value.replace(/\./g, '') : value)
  } else {
    candidates.push(value)
  }
  const last = value.match(/([.,])(\d{2})$/)
  if (last?.index !== undefined) {
    const intPart = value.slice(0, last.index).replace(/[^0-9]/g, '')
    if (intPart) candidates.push(`${intPart}.${last[2]}`)
  }
  for (const candidate of candidates) {
    const parsed = Number(candidate)
    if (!Number.isNaN(parsed)) return parsed.toFixed(2).replace('.', ',')
  }
  return 'NA'
}

export function extractInvoiceNumber(text: string): string {
  const patterns = [
    /\bEingangsrechnung\s+([A-Z0-9][A-Z0-9./-]{2,})\b/i,
    /\bRechnungs\s*nummer\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b/i,
    /\bRechnungs\s*nr\.?\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b/i,
    /\bRechnungs\s*-\s*Nr\.?\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b/i,
    /\bRechnung\s*-\s*Nr\.?\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b/i,
    /\bRechnung\s*Nr\.?\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b/i,
    /\bInvoice\s*No\.?\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b/i,
    /\bInvoice\s*Number\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b/i,
    /\bBeleg(?:nr|nummer)\.?\s*[:#.]?\s*([A-Z0-9][A-Z0-9./-]{1,})\b/i
  ]
  for (const line of text.split(/\r?\n/).map((x) => x.trim()).filter(Boolean)) {
    if (/kundennummer|customer|auftrags|bestell/i.test(line)) continue
    for (const pattern of patterns) {
      const match = line.match(pattern)
      if (match) return match[1].replace(/[^A-Za-z0-9./-]+/g, '').slice(0, 40)
    }
  }
  return 'NA'
}

function invoiceDate(text: string, createdAt?: Date): string {
  const lines = text.split(/\r?\n/).map((x) => x.trim()).filter(Boolean)
  for (const line of lines) if (/rechnungsdatum|invoice date/i.test(line)) return normalizeDate(line) || '00/00/0000'
  for (const line of lines) {
    if (/lieferdatum|leistungsdatum|valutadatum/i.test(line)) continue
    if (/\bdatum\b|\bdate\b/i.test(line)) return normalizeDate(line) || '00/00/0000'
  }
  return createdAt ? `${createdAt.getDate().toString().padStart(2, '0')}/${(createdAt.getMonth() + 1).toString().padStart(2, '0')}/${createdAt.getFullYear()}` : '00/00/0000'
}

function invoiceAmount(text: string): string {
  const nums = /(?<!\d)(\d{1,3}(?:[.,]\d{3})*(?:[.,]\d{2})|\d+[.,]\d{2}|\d+)(?!\d)/g
  const strong = /endsumme|gesamtsumme|gesamtbetrag|zu zahlen|balance due|invoice total|grand total|brutto/i
  const medium = /summe|gesamt|total|rechnungsbetrag/i
  const bad = /netto|mwst|ust|steuer|skonto|rabatt/i
  const rows: Array<[number, number, string]> = []
  for (const line of text.split(/\r?\n/)) {
    const found = [...line.matchAll(nums)].map((x) => x[1])
    if (!found.length) continue
    let score = strong.test(line) ? 4 : medium.test(line) ? 2 : 0
    if (bad.test(line) && !strong.test(line)) score -= 3
    const amount = normalizeAmount(found.at(-1) ?? '')
    if (amount !== 'NA') rows.push([score, Number(amount.replace(',', '.')), amount])
  }
  rows.sort((a, b) => b[0] - a[0] || b[1] - a[1])
  return rows[0]?.[2] ?? 'NA'
}

export function extractBelegeSender(text: string, originalFilename = ''): string {
  const lines = text.split(/\r?\n/).map((x) => x.trim()).filter(Boolean).slice(0, 20)
  if (/^[A-Z]{3,8}$/.test(lines[0] ?? '') && /^[A-Z][A-Z ]{3,24}$/.test(lines[1] ?? '')) return compactSender(lines[0])
  for (const line of lines.slice(0, 12)) {
    const token = compactSender(line)
    const low = token.toLowerCase()
    if (!token || badBelege.some((x) => low.includes(x)) || /^(\d+|rechnung|invoice|betrag|datum|summe)$/i.test(token)) continue
    if (line.includes(':') || /\d{2,}/.test(line)) continue
    if (/(bank|gmbh|ag|market|shop|store|distributors)/i.test(line) || /^[A-Z][A-Za-z0-9]{2,20}$/.test(token)) return token
  }
  const stem = originalFilename.replace(/\.[A-Za-z0-9]{1,5}$/, '')
  if (stem && !/(scan|img|image|document|invoice|rechnung|impf|therapie)/i.test(stem)) return compactSender(stem)
  return 'Dok'
}

function belegeMonthYear(text: string, createdAt?: Date): string {
  const date = normalizeDate(text)
  if (date) return `${date.slice(3, 5)}/${date.slice(-2)}`
  return createdAt ? `${(createdAt.getMonth() + 1).toString().padStart(2, '0')}/${(createdAt.getFullYear() % 100).toString().padStart(2, '0')}` : '00/00'
}

function belegeAmount(text: string): string {
  const rows: Array<[number, string]> = []
  const nums = /(?<!\d)(\d{1,3}(?:[.,]\d{3})*(?:[.,]\d{2})|\d+[.,]\d{2}|\d+)(?!\d)/g
  for (const line of text.split(/\r?\n/)) {
    let score = /(summe|gesamt|total|betrag|balance due|zu zahlen|endbetrag|invoice total|rechnungsbetrag)/i.test(line) ? 2 : 0
    if (/eur|€|\$/i.test(line)) score += 1
    for (const match of line.matchAll(nums)) rows.push([score, match[1]])
  }
  rows.sort((a, b) => b[0] - a[0] || b[1].length - a[1].length)
  return rows[0]?.[0] >= 2 ? normalizeAmount(rows[0][1]) : 'NA'
}

function payment(text: string): string {
  if (/\b(bar|cash|barzahlung)\b/i.test(text)) return 'Bar'
  if (/\b(karte|ec|girocard|visa|mastercard|amex|electronic cash|debit)\b/i.test(text)) return 'Karte'
  return 'NA'
}

export function extractEingangsrechnungSender(text: string, originalFilename = ''): string {
  for (const line of text.split(/\r?\n/).map((x) => x.trim()).filter(Boolean).slice(0, 18)) {
    if (/(firma|kunde|rechnung|invoice|datum|summe|telefon|mail|www|lieferdatum)/i.test(line)) continue
    const candidate = line.split(/\b(phone|telefon|fax|mail|www)\b/i)[0].split(/\b(strasse|straße|str\.?|weg|gasse|allee|platz|ufer|ring)\b/i)[0].split(/\b\d{4,5}\b/)[0]
    const sender = compactParty(candidate)
    if (sender !== 'Dok' && sender.length >= 3) return sender
  }
  return originalFilename ? compactParty(originalFilename.replace(/\.[A-Za-z0-9]{1,5}$/, '')) : 'Dok'
}

export function extractAusgangsrechnungRecipient(text: string): string {
  const meta = /\b(rechnung|rechnungs|invoice|datum|date|kunden-?nr|kundennummer|bestell-?nr|auftrags-?nr|telefon|phone|fax|mail|www|summe|ust|mwst|endsumme|iban|bic|konto|pos\.)/i
  const top: string[] = []
  for (const line of text.split(/\r?\n/).map((x) => x.trim()).filter(Boolean).slice(0, 25)) {
    if (meta.test(line)) break
    top.push(line)
  }
  const clean = top
    .filter((line) => !meta.test(line) && !/@|www\.|\+\d/.test(line) && !/^\d{4,5}\s/.test(line) && !/\d/.test(line) && !/\b(strasse|straße|str\.?|weg|gasse|allee|platz|ufer|ring)\b/i.test(line))
    .map(compactParty)
    .filter((x, i, rows) => x !== 'Dok' && x.length >= 3 && rows.indexOf(x) === i)
  return clean[1] ?? 'Dok'
}

export function extractBelegeTitle(input: ExtractionInput): string {
  return `${extractBelegeSender(input.text, input.originalFilename)}_B_${belegeMonthYear(input.text, input.createdAt)}_${belegeAmount(input.text)}_${payment(input.text)}`
}

export function extractEingangsrechnungTitle(input: ExtractionInput): string {
  return `${extractEingangsrechnungSender(input.text, input.originalFilename)}_${extractInvoiceNumber(input.text)}_${invoiceDate(input.text, input.createdAt)}_${invoiceAmount(input.text)}`
}

export function extractAusgangsrechnungTitle(input: ExtractionInput): string {
  return `${extractAusgangsrechnungRecipient(input.text)}_${extractInvoiceNumber(input.text)}_${invoiceDate(input.text, input.createdAt)}_${invoiceAmount(input.text)}`
}
